10_rest/runnerdave/pkg/puppy/rest_test.go

```diff
@@ -0,0 +1,224 @@
+package puppy_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	puppy "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy"
+	store "github.com/anz-bank/go-course/10_rest/runnerdave/pkg/puppy/store"
+)
+
+var (
+	rs puppy.RestServer
+)
+
+func TestGetPuppy(t *testing.T) {
+	req, err := http.NewRequest("GET", "/api/puppy/1", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rr := httptest.NewRecorder()
+	newPuppy := puppy.Puppy{
+		Breed:  "ULTRASURE",
+		Colour: "green",
+		Value:  2732.81,
+	}
+	db := store.NewSyncStore()
+	if err := db.CreatePuppy(newPuppy); err != nil {
+		panic(err)
+	}
+	handler := rs.SetupRoutes(db)
+
+	handler.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusOK {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusOK)
+	}
+
+	expected := `{
+		"id": 1,
+		"breed": "ULTRASURE",
+		"color": "green",
+		"value": 2732.81
+	  }`
+	actual := rr.Body.String()
+	require.JSONEq(t, expected, actual, "GET Puppy was incorrect, got: %s, want: %s.",
```

Comment:
I would leave out the error message here. All the information you put in this message will be written out automatically - the name of the test that failed, and the actual vs expected values. It makes it much shorter then. I'd also move the previous assertion down to here too, so they are together, and get rid of the `actual` variable:
```
assert.Equal(t, http.StatusOK, rr.Code)
assert.JSONEq(t, expected, rr.Body.String())
```

By getting rid of those intermediate variables, these assertions are very clear what is being tested without needing to look at any other lines of code. Once someone becomes familiar with the `httptest.ResponseRecorder` type, this reads very naturally.

Please structure your other tests the same way.

Reply:
fixed across the file